### PR TITLE
Fix two O(n^2) performance bugs in the sysctl.present state

### PR DESCRIPTION
**Pull request by asomers**
### What does this PR do?
Fixes two O(n^2) performance bugs in the sysctl.present state

### Previous Behavior
Two bugs:
* When `test=True`, the `sysctl.present` state would call the sysctl module to get the values of _every_single_sysctl_.  Sometimes there are tens of thousands of those.  And it would do that again, for _every_single_ `sysctl.present` state.
* The `sysctl.show` function would split the output of the `sysctl` command line-by-line.  For multiline values, it would reassemble them by appending to a string, which is `O(n^2)`.

### New Behavior
* Now the `sysctl.present` state will call `sysctl.get` instead of `sysctl.show`, only looking up a single value.
* `sysctl.show` concatenates strings using `str.join`, which is much faster.

### Merge requirements satisfied?
- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

### Tests run?
yes

### CODE REVIEW

**Comment by asomers** [comment]
Fixing the tests is blocked by #58746 .

File: salt/states/sysctl.py
Comment (Ch3LL) [inline salt/states/sysctl.py:55]: Adding name to this dictionary will lead to this line `if name in current and name not in configured:` always passing the first condition, and we only want to pass that condition if that name actually exists.

**Review by Ch3LL — CHANGES_REQUESTED** [review]
Can you also add a changelog entry as documented here: https://docs.saltstack.com/en/master/topics/development/changelog.html

**Comment by waynew** [comment]
Hey @asomers thanks for the PR!

I've been OOO for a bit and am just catching up, apologies for the wait! I'll be reviewing this PR as soon as I can, thanks for your patience!

**Comment by asomers** [comment]
Thanks @waynew .  But please review https://github.com/saltstack/salt/pull/58746 first.  That PR blocks this one.

File: salt/states/sysctl.py
Comment (asomers) [reply salt/states/sysctl.py]: Fixed

**Comment by asomers** [comment]
I rebased to incorporate the changes from PR #58746 , and I also fixed @Ch3LL's objection.  This should be good to review now.

**Comment by asomers** [comment]
On one particular server that I manage, this change reduces the time for `state.apply test=True` from 126 minutes down to less than 6 minutes.  On another server, I previously timed that operation as taking more than 5 hours, but I haven't rerun it with this patch yet.

**Comment by asomers** [comment]
I added the changelog entry as @Ch3LL requested.

**Review by krionbsd — DISMISSED** [review]

File: tests/unit/modules/test_freebsd_sysctl.py
Comment (dwoz) [inline tests/unit/modules/test_freebsd_sysctl.py:128]: Please use `tests.supper.helpers.wrap` for `return_value`

**Review by dwoz — CHANGES_REQUESTED** [review]

File: tests/unit/modules/test_freebsd_sysctl.py
Comment (waynew) [reply tests/unit/modules/test_freebsd_sysctl.py:128]: Was that `tests.support.helpers.wrap`?  Though I'm not seeing that :thinking: - or are you referring to textwrap?

File: salt/states/sysctl.py
Comment (waynew) [inline salt/states/sysctl.py:89]: **question**: As I look at this `if/elif` chain would it be possible to rewrite this a bit clearer, perhaps by shuffling the order/indenting things?

File: salt/states/sysctl.py
Comment (waynew) [inline salt/states/sysctl.py:65]: **suggestion**: because `""` is falsey this could be rewritten as:
```suggestion
        if current and name not in configured:
```

**Review by waynew — CHANGES_REQUESTED** [review]
Looks good, just a couple of minor (potential) improvements.

File: tests/unit/modules/test_freebsd_sysctl.py
Comment (dwoz) [reply tests/unit/modules/test_freebsd_sysctl.py:128]: Sorry `tests.support.helpers.dedent`

File: tests/unit/states/test_sysctl.py
Comment (asomers) [reply tests/unit/states/test_sysctl.py:176]: I can't use create_autospec, because there is no `sysctl.get` object.  And the `sysctl.__salt__` dict is empty until I put a mock inside.

File: salt/states/sysctl.py
Comment (asomers) [reply salt/states/sysctl.py:89]: Ok.  I made the chain a little deeper but less long now.

File: tests/unit/states/test_sysctl.py
Comment (waynew) [reply tests/unit/states/test_sysctl.py:176]: 🤔 if there's no sysctl.get, what is called with `__salt__['sysctl.get']`?

Comment (asomers) [reply tests/unit/states/test_sysctl.py:176]: The mock object that gets put there by `patch.dict`

Comment (waynew) [reply tests/unit/states/test_sysctl.py:176]: Right, but normally it exists /somewhere/, right? Or the code wouldn't work when calling via the command line.

That's the function to autospec

Comment (asomers) [reply tests/unit/states/test_sysctl.py:176]: I'm sure that it normally exists somewhere.  But I can't figure out where.  Does it have something to do with `__virtualname__` and `__virtual__()`?  Are modules like these not loaded during the state unit tests?

**Comment by asomers** [comment]
The test failures look unrelated.

File: tests/unit/states/test_sysctl.py
Comment (waynew) [reply tests/unit/states/test_sysctl.py:176]: Ah, that's where the confusion is :upside_down_face: 

Normally `__salt__` will contain a `'module.function'` for everything in `salt/modules/`, but `def __virtual__():` does allow you to change the name. This is useful for cases where you need different behavior based on operating system or even underlying library.

In /this/ case, there are several files in `modules/*_sysctl.py` that could apply. Fortunately they all have the same API, so any one of them could apply. I'm not sure if there's a better way for `create_autospec`, though :thinking:

Comment (asomers) [reply tests/unit/states/test_sysctl.py:176]: I guessed it must be something like that.  But in this case, there's literally nothing in `sysctl.__salt__`:
```
(Pdb) p sysctl.__salt__
{}
```

Comment (waynew) [reply tests/unit/states/test_sysctl.py:176]: Yeah - that's because within the (unit) test suites we avoid loading all the things :upside_down_face:

**Comment by asomers** [comment]
@waynew there's still one test failure, but it looks unrelated to this PR.  What is the next step?

**Review by waynew — APPROVED** [review]
@asomers good question. Hopefully this will be fixed in the main branch - pinged the core team to ask. I'm approving it but we may need to rebase/merge `master` to get that test green (or re-run the suite if it's an infrastructure issue)

**Comment by asomers** [comment]
Thanks Wayne.  Just ping me if I need to do anything.

**Review by Ch3LL — APPROVED** [review]

**Comment by asomers** [comment]
@waynew @Ch3LL is there anything else we need to do there?

**Comment by Ch3LL** [comment]
Should be good, just need to get the tests green. A lot of cleanup on pytest has occurred recently so i'll merge to master and should be good to go after that.

**Pull request merged**
